common/flatpak-utils.c

```diff
@@ -2552,69 +2552,119 @@ flatpak_variant_bsearch_str (GVariant   *array,
+                                               G_VARIANT_TYPE ("a{sa(s(taya{sv}))}"));
+      if (collection_map != NULL)
+        refs = g_variant_lookup_value (collection_map, collection_id, G_VARIANT_TYPE ("a(s(taya{sv}))"));
+    }
+
+  return g_steal_pointer (&refs);
+}
+
+/* This matches all refs from @collection_id that have ref, followed by '.'  as prefix */
 char **
-flatpak_summary_match_subrefs (GVariant *summary, const char *ref)
+flatpak_summary_match_subrefs (GVariant   *summary,
+                               const char *collection_id,
+                               const char *ref)
 {
-  g_autoptr(GVariant) refs = g_variant_get_child_value (summary, 0);
+  g_autoptr(GVariant) refs = NULL;
+  g_autoptr(GVariant) metadata = g_variant_get_child_value (summary, 1);
   GPtrArray *res = g_ptr_array_new ();
   gsize n, i;
   g_auto(GStrv) parts = NULL;
   g_autofree char *parts_prefix = NULL;
   g_autofree char *ref_prefix = NULL;
   g_autofree char *ref_suffix = NULL;
 
-  parts = g_strsplit (ref, "/", 0);
-  parts_prefix = g_strconcat (parts[1], ".", NULL);
-
-  ref_prefix = g_strconcat (parts[0], "/", NULL);
-  ref_suffix = g_strconcat ("/", parts[2], "/", parts[3], NULL);
+  /* Work out which refs list to use, based on the @collection_id. */
+  refs = summary_find_refs_list (summary, collection_id);
 
-  n = g_variant_n_children (refs);
-  for (i = 0; i < n; i++)
+  if (refs != NULL)
     {
-      g_autoptr(GVariant) child = NULL;
-      g_autoptr(GVariant) cur_v = NULL;
-      const char *cur;
-      const char *id_start;
+      /* Match against the refs. */
+      parts = g_strsplit (ref, "/", 0);
+      parts_prefix = g_strconcat (parts[1], ".", NULL);
 
-      child = g_variant_get_child_value (refs, i);
-      cur_v = g_variant_get_child_value (child, 0);
-      cur = g_variant_get_data (cur_v);
+      ref_prefix = g_strconcat (parts[0], "/", NULL);
+      ref_suffix = g_strconcat ("/", parts[2], "/", parts[3], NULL);
 
-      /* Must match type */
-      if (!g_str_has_prefix (cur, ref_prefix))
-        continue;
+      n = g_variant_n_children (refs);
+      for (i = 0; i < n; i++)
+        {
+          g_autoptr(GVariant) child = NULL;
+          g_autoptr(GVariant) cur_v = NULL;
+          const char *cur;
+          const char *id_start;
 
-      /* Must match arch & branch */
-      if (!g_str_has_suffix (cur, ref_suffix))
-        continue;
+          child = g_variant_get_child_value (refs, i);
+          cur_v = g_variant_get_child_value (child, 0);
+          cur = g_variant_get_data (cur_v);
 
-      id_start = strchr (cur, '/');
-      if (id_start == NULL)
-        continue;
+          /* Must match type */
+          if (!g_str_has_prefix (cur, ref_prefix))
+            continue;
 
-      /* But only prefix of id */
-      if (!g_str_has_prefix (id_start + 1, parts_prefix))
-        continue;
+          /* Must match arch & branch */
+          if (!g_str_has_prefix (cur, ref_suffix))
+            continue;
```

Comment:
Review comment from @alexlarsson on a plane: This reverts an existing fix (bad merge conflict resolution).